### PR TITLE
feat(risedev): use jsonnet for profiles and templates

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

[Jsonnet](https://jsonnet.org/) is a powerful tool to write reusable and extensible configurations.

In this PR, we introduced jsonnet-based RiseDev profiles and templates. Benefits are:

- Get rid of the Rust "expanders" for manipulating profiles. All transformations are done in jsonnet.
- Be able to reuse the service list between different profiles.
- Code-completion and hints when writing new profiles or service templates.

The compatibility for old `risedev.yml` format is still kept through `compat.libsonnet`. We can migrate them in the future.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: risedev-profiles.libsonnet
Comment (BugenZhao) [inline risedev-profiles.libsonnet:1]: This is where future risedev profiles will reside.

File: risedev-template.libsonnet
Comment (BugenZhao) [inline risedev-template.libsonnet:1]: Migrated from the `template` section in `risedev.yml`.

File: risedev-template.libsonnet [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+{
+  minio: {
+    // Id of this instance
+    id: 'minio',
+    // Advertise address of MinIO s3 endpoint
+    address: '127.0.0.1',
+    // Advertise port of MinIO s3 endpoint
+    port: 9301,
+    // Listen address of MinIO endpoint
+    listenAddress: self.address,
```
Comment (BugenZhao) [inline risedev-template.libsonnet:12]: No need to use `DollarExpander` anymore.

File: risedev-template.libsonnet [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+    // '0' means this env var will use the default of minio.
+    apiRequestsMax: 0,
+    // Deadline for api requests.
+    // Empty string means this env var will use the default of minio.
+    apiRequestsDeadline: '',
+  },
+
+  etcd: {
+    // Id of this instance
+    id: 'etcd-' + self.port,
```
Comment (BugenZhao) [inline risedev-template.libsonnet:36]: No need to use `IdExpander` anymore.

File: src/risedevtool/jsonnet/expand.libsonnet [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  { use: std.splitLimitR(step.id, '-', 1)[0] }
+  // Expand all `provide-xx` fields.
+  {
+    [name]:
+      [
+        // Expand the provided steps.
+        // The nested `provide-xx` fields will not be used. Prune them.
+        pruneProvide(otherStep)
+        for otherStep in steps
+        if std.startsWith(otherStep.id, std.rstripChars(step[name], '*'))
```
Comment (BugenZhao) [inline src/risedevtool/jsonnet/expand.libsonnet:25]: No need for `ProvideExpander` anymore.

File: src/risedevtool/jsonnet/main.jsonnet
Comment (BugenZhao) [inline src/risedevtool/jsonnet/main.jsonnet:1]: This is the entrypoint of the jsonnet program to evaluate.

File: src/risedevtool/jsonnet/profiles-compat.libsonnet
Comment (BugenZhao) [inline src/risedevtool/jsonnet/profiles-compat.libsonnet:1]: This is to convert `risedev.yml` to the format of `risedev-profiles.libsonnet`.

File: src/risedevtool/src/bin/risedev-dev.rs [hunk 4]
```diff
@@ -336,6 +337,13 @@ fn task_main(
     Ok((stat, log_buffer))
 }
 
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+struct Input {
+    config_path: Option<String>,
+    steps: Vec<ServiceConfig>,
+}
```
Comment (BugenZhao) [inline src/risedevtool/src/bin/risedev-dev.rs:343]: The generated JSON will be directly deserialized into this struct, then executed by `risedev-dev`.

**Review by xxchan — APPROVED** [review]
Generally LGTM (not delved into the details), although I don't see there's any compelling reasons to do the refactoring (risedev config expander works good enough).

Besides, please update `src/risedevtool/README.md`. `src/risedevtool/risedev-schema.json` may also need update.

**Review by stdrc — APPROVED** [review]
LGTM since it seems to do no harm to current dev experience and at the same time delete so many LoC

**Pull request closed without merging**